File: backend/utils/device_fingerprint.py
import hashlib
import json
from flask import request
from user_agents import parse

def generate_device_fingerprint(user_agent=None, accept_language=None, additional_data=None):
    """
    Generate a device fingerprint from browser characteristics
    This creates a semi-persistent device ID that survives browser restarts
    but changes if user clears all data or uses incognito mode extensively
    """
    if not user_agent:
        user_agent = request.headers.get('User-Agent', '')
    
    if not accept_language:
        accept_language = request.headers.get('Accept-Language', '')
    
    # Parse user agent for detailed info
    parsed_ua = parse(user_agent)
    
    # Core fingerprint components
    fingerprint_data = {
        'browser_family': parsed_ua.browser.family,
        'browser_version': f"{parsed_ua.browser.version[0]}.{parsed_ua.browser.version[1]}" if len(parsed_ua.browser.version) >= 2 else str(parsed_ua.browser.version[0]) if parsed_ua.browser.version else 'unknown',
        'os_family': parsed_ua.os.family,
        'os_version': f"{parsed_ua.os.version[0]}.{parsed_ua.os.version[1]}" if len(parsed_ua.os.version) >= 2 else str(parsed_ua.os.version[0]) if parsed_ua.os.version else 'unknown',
        'device_family': parsed_ua.device.family,
        'is_mobile': parsed_ua.is_mobile,
        'is_tablet': parsed_ua.is_tablet,
        'is_pc': parsed_ua.is_pc,
        'accept_language': accept_language[:10] if accept_language else '',  # First 10 chars to avoid too much specificity
    }
    
    # Add any additional data provided by frontend
    if additional_data:
        # Screen resolution, timezone, etc. from frontend JavaScript
        if isinstance(additional_data, dict):
            fingerprint_data.update({
                'screen_width': additional_data.get('screen_width'),
                'screen_height': additional_data.get('screen_height'),
                'timezone': additional_data.get('timezone'),
                'color_depth': additional_data.get('color_depth'),
                'pixel_ratio': additional_data.get('pixel_ratio'),
            })
    
    # Create deterministic hash
    fingerprint_string = json.dumps(fingerprint_data, sort_keys=True)
    device_id = hashlib.sha256(fingerprint_string.encode()).hexdigest()[:32]
    
    return device_id, fingerprint_data

def get_device_info(user_agent=None, ip_address=None):
    """
    Extract human-readable device information
    """
    if not user_agent:
        user_agent = request.headers.get('User-Agent', '')
    
    if not ip_address:
        ip_address = get_client_ip()
    
    parsed_ua = parse(user_agent)
    
    # Generate friendly device name
    device_name = f"{parsed_ua.browser.family}"
    if parsed_ua.os.family and parsed_ua.os.family != 'Other':
        device_name += f" on {parsed_ua.os.family}"
    
    if parsed_ua.device.family and parsed_ua.device.family != 'Other':
        device_name = f"{parsed_ua.device.family} - {device_name}"
    
    return {
        'name': device_name,
        'browser': {
            'name': parsed_ua.browser.family,
            'version': str(parsed_ua.browser.version) if parsed_ua.browser.version else 'Unknown'
        },
        'os': {
            'name': parsed_ua.os.family,
            'version': str(parsed_ua.os.version) if parsed_ua.os.version else 'Unknown'
        },
        'device': {
            'family': parsed_ua.device.family,
            'brand': parsed_ua.device.brand,
            'model': parsed_ua.device.model
        },
        'type': 'mobile' if parsed_ua.is_mobile else 'tablet' if parsed_ua.is_tablet else 'desktop',
        'is_mobile': parsed_ua.is_mobile,
        'is_tablet': parsed_ua.is_tablet,
        'is_pc': parsed_ua.is_pc,
        'ip_address': ip_address
    }

def get_client_ip():
    """
    Get the real client IP address, handling proxies and load balancers
    """
    # Check for forwarded IP first (from load balancers, proxies)
    forwarded_for = request.headers.get('X-Forwarded-For')
    if forwarded_for:
        # X-Forwarded-For can contain multiple IPs, client IP is first
        return forwarded_for.split(',')[0].strip()
    
    # Check other common headers
    real_ip = request.headers.get('X-Real-IP')
    if real_ip:
        return real_ip
    
    # Fall back to direct connection
    return request.remote_addr

def generate_frontend_fingerprint_script():
    """
    Generate JavaScript code for frontend fingerprinting
    This collects additional browser characteristics that aren't available server-side
    """
    return """
    function generateDeviceFingerprint() {
        const canvas = document.createElement('canvas');
        const ctx = canvas.getContext('2d');
        ctx.textBaseline = 'top';
        ctx.font = '14px Arial';
        ctx.fillText('Device fingerprint', 2, 2);
        
        return {
            screen_width: screen.width,
            screen_height: screen.height,
            available_width: screen.availWidth,
            available_height: screen.availHeight,
            color_depth: screen.colorDepth,
            pixel_ratio: window.devicePixelRatio,
            timezone: Intl.DateTimeFormat().resolvedOptions().timeZone,
            language: navigator.language,
            languages: navigator.languages ? navigator.languages.join(',') : '',
            platform: navigator.platform,
            cookie_enabled: navigator.cookieEnabled,
            do_not_track: navigator.doNotTrack,
            canvas_fingerprint: canvas.toDataURL(),
            local_storage: !!window.localStorage,
            session_storage: !!window.sessionStorage,
            indexed_db: !!window.indexedDB,
            web_gl_vendor: (() => {
                try {
                    const canvas = document.createElement('canvas');
                    const gl = canvas.getContext('webgl') || canvas.getContext('experimental-webgl');
                    return gl ? gl.getParameter(gl.VENDOR) : null;
                } catch (e) {
                    return null;
                }
            })(),
            touch_support: 'ontouchstart' in window,
            max_touch_points: navigator.maxTouchPoints || 0
        };
    }
    
    // Make available globally
    window.generateDeviceFingerprint = generateDeviceFingerprint;
    """

def create_device_session_token(device_id, user_id, session_data=None):
    """
    Create a device-specific session token that can be used for
    additional security verification
    """
    import time
    
    token_data = {
        'device_id': device_id,
        'user_id': user_id,
        'timestamp': int(time.time()),
        'session_data': session_data or {}
    }
    
    token_string = json.dumps(token_data, sort_keys=True)
    return hashlib.sha256(token_string.encode()).hexdigest()[:16]

def is_device_fingerprint_similar(fingerprint1, fingerprint2, similarity_threshold=0.8):
    """
    Check if two device fingerprints are similar enough to be considered the same device
    Useful for handling minor browser updates or setting changes
    """
    if not fingerprint1 or not fingerprint2:
        return False
    
    # Compare key components
    key_components = [
        'browser_family', 'os_family', 'device_family', 
        'is_mobile', 'is_tablet', 'is_pc'
    ]
    
    matches = 0
    total = len(key_components)
    
    for component in key_components:
        if fingerprint1.get(component) == fingerprint2.get(component):
            matches += 1
    
    similarity = matches / total
    return similarity >= similarity_threshold

File: frontend/src/lib/deviceFingerprint.ts
/**
 * Device Fingerprinting Utility
 * Generates a semi-persistent device identifier for authentication
 */

export interface DeviceFingerprint {
  screen_width: number;
  screen_height: number;
  available_width: number;
  available_height: number;
  color_depth: number;
  pixel_ratio: number;
  timezone: string;
  language: string;
  languages: string;
  platform: string;
  cookie_enabled: boolean;
  do_not_track: string | null;
  canvas_fingerprint: string;
  local_storage: boolean;
  session_storage: boolean;
  indexed_db: boolean;
  web_gl_vendor: string | null;
  touch_support: boolean;
  max_touch_points: number;
}

export function generateDeviceFingerprint(): DeviceFingerprint {
  const canvas = document.createElement('canvas');
  const ctx = canvas.getContext('2d');
  if (ctx) {
    ctx.textBaseline = 'top';
    ctx.font = '14px Arial';
    ctx.fillText('Device fingerprint', 2, 2);
  }
  
  return {
    screen_width: screen.width,
    screen_height: screen.height,
    available_width: screen.availWidth,
    available_height: screen.availHeight,
    color_depth: screen.colorDepth,
    pixel_ratio: window.devicePixelRatio,
    timezone: Intl.DateTimeFormat().resolvedOptions().timeZone,
    language: navigator.language,
    languages: navigator.languages ? navigator.languages.join(',') : '',
    platform: navigator.platform,
    cookie_enabled: navigator.cookieEnabled,
    do_not_track: navigator.doNotTrack,
    canvas_fingerprint: canvas.toDataURL(),
    local_storage: !!window.localStorage,
    session_storage: !!window.sessionStorage,
    indexed_db: !!window.indexedDB,
    web_gl_vendor: getWebGLVendor(),
    touch_support: 'ontouchstart' in window,
    max_touch_points: navigator.maxTouchPoints || 0
  };
}

function getWebGLVendor(): string | null {
  try {
    const canvas = document.createElement('canvas');
    const gl = canvas.getContext('webgl') || canvas.getContext('experimental-webgl');
    if (gl && 'getParameter' in gl) {
      const webgl = gl as WebGLRenderingContext;
      const debugInfo = webgl.getExtension('WEBGL_debug_renderer_info');
      if (debugInfo) {
        return webgl.getParameter(debugInfo.UNMASKED_VENDOR_WEBGL);
      }
      return webgl.getParameter(webgl.VENDOR);
    }
    return null;
  } catch (e) {
    return null;
  }
}

/**
 * Get or create a persistent device ID from localStorage
 * Falls back to session-based ID if localStorage is not available
 */
export function getDeviceId(): string {
  const DEVICE_ID_KEY = 'intralink_device_id';
  
  try {
    // Try to get existing device ID
    let deviceId = localStorage.getItem(DEVICE_ID_KEY);
    
    if (!deviceId) {
      // Generate new device ID
      deviceId = generateSecureDeviceId();
      localStorage.setItem(DEVICE_ID_KEY, deviceId);
    }
    
    return deviceId;
  } catch (e) {
    // localStorage not available, use session storage or generate temporary ID
    try {
      let deviceId = sessionStorage.getItem(DEVICE_ID_KEY);
      if (!deviceId) {
        deviceId = generateSecureDeviceId();
        sessionStorage.setItem(DEVICE_ID_KEY, deviceId);
      }
      return deviceId;
    } catch (e) {
      // Generate temporary device ID
      return generateSecureDeviceId();
    }
  }
}

function generateSecureDeviceId(): string {
  const timestamp = Date.now().toString();
  const random = Math.random().toString(36).substr(2, 9);
  const fingerprint = generateBasicFingerprint();
  
  // Combine and hash
  const combined = `${timestamp}-${random}-${fingerprint}`;
  return btoa(combined).replace(/[+/=]/g, '').substr(0, 32);
}

function generateBasicFingerprint(): string {
  const components = [
    navigator.userAgent,
    navigator.language,
    screen.width,
    screen.height,
    screen.colorDepth,
    new Date().getTimezoneOffset()
  ];
  
  return components.join('|');
}

/**
 * Check if device supports persistent storage
 */
export function supportsPeristentStorage(): boolean {
  try {
    return !!window.localStorage;
  } catch (e) {
    return false;
  }
}

/**
 * Get device information for display purposes
 */
export function getDeviceInfo(): {
  type: 'mobile' | 'tablet' | 'desktop';
  browser: string;
  os: string;
  name: string;
} {
  const userAgent = navigator.userAgent;
  
  // Detect device type
  let type: 'mobile' | 'tablet' | 'desktop' = 'desktop';
  if (/tablet|ipad/i.test(userAgent)) {
    type = 'tablet';
  } else if (/mobile|android|iphone|ipod|blackberry|windows phone/i.test(userAgent)) {
    type = 'mobile';
  }
  
  // Detect browser
  let browser = 'Unknown';
  if (userAgent.includes('Chrome')) browser = 'Chrome';
  else if (userAgent.includes('Firefox')) browser = 'Firefox';
  else if (userAgent.includes('Safari')) browser = 'Safari';
  else if (userAgent.includes('Edge')) browser = 'Edge';
  else if (userAgent.includes('Opera')) browser = 'Opera';
  
  // Detect OS
  let os = 'Unknown';
  if (userAgent.includes('Windows')) os = 'Windows';
  else if (userAgent.includes('Mac')) os = 'macOS';
  else if (userAgent.includes('Linux')) os = 'Linux';
  else if (userAgent.includes('Android')) os = 'Android';
  else if (userAgent.includes('iOS')) os = 'iOS';
  
  // Generate friendly name
  const name = `${browser} on ${os}`;
  
  return { type, browser, os, name };
}

/**
 * Clear device ID (for logout/reset)
 */
export function clearDeviceId(): void {
  const DEVICE_ID_KEY = 'intralink_device_id';
  
  try {
    localStorage.removeItem(DEVICE_ID_KEY);
  } catch (e) {
    // Ignore
  }
  
  try {
    sessionStorage.removeItem(DEVICE_ID_KEY);
  } catch (e) {
    // Ignore
  }
}